src/plugins/intel_gpu/src/graph/impls/cpu/non_max_suppression.cpp

```diff
@@ -440,6 +440,121 @@ attach_non_max_suppression_impl::attach_non_max_suppression_impl() {
+    auto dep_mem_batch = static_cast<size_t>(dep_mem_layout.batch());
+
+    mem_lock<T, mem_lock_type::read> dep_mem_lock(mem, stream);
+    auto dep_mem_ptr = dep_mem_lock.data();
+
+    size_t actual_valid_num = dep_mem_batch;
+    size_t idx = 0;
+    for (size_t i = 0; i < dep_mem_batch; i++) {
+        idx = i * 3;
+        if (dep_mem_ptr[idx] == -1) {
+            actual_valid_num = i;
+            break;
+        }
+    }
+
+    return actual_valid_num;
+}
+
+template <typename T>
+void store_nms_gather_output(non_max_suppression_gather_inst& instance, size_t idx, size_t valid_size) {
+    auto input_mem = instance.dep_memory_ptr(idx);
+    layout dep_layout = input_mem->get_layout();
+    auto output_ps = dep_layout.get_partial_shape();
+
+    output_ps[0] = valid_size;   // update valid batch size
+    auto output_layout = layout(output_ps, dep_layout.data_type, dep_layout.format);
+    auto new_output_mem = instance.get_network().get_engine().reinterpret_buffer(*input_mem, output_layout);
+
+    instance.set_output_memory(new_output_mem, true, idx);
+}
+
+void run_nms_gather(non_max_suppression_gather_inst& instance) {
+    auto& stream = instance.get_network().get_stream();
+
+    auto valid_input_batch = get_nms_gather_valid_size<ov::element_type_traits<data_types::i32>::value_type>(stream, instance.dep_memory_ptr(0));
+    store_nms_gather_output<ov::element_type_traits<data_types::i32>::value_type>(instance, 0, valid_input_batch);
+
+    if (instance.outputs_memory_count() >= 2) {
+        auto data_type = instance.dep_memory_ptr(1)->get_layout().data_type;
+
+        if (data_type == cldnn::data_types::f16) {
+            store_nms_gather_output<ov::element_type_traits<data_types::f16>::value_type>(instance, 1, valid_input_batch);
+        } else if (data_type == cldnn::data_types::f32) {
+            store_nms_gather_output<ov::element_type_traits<data_types::f32>::value_type>(instance, 1, valid_input_batch);
+        } else {
+            throw std::runtime_error("Non max suppression gather - unsupported second output data type");
+        }
+
+        if (instance.outputs_memory_count() == 3) {
+            mem_lock<ov::element_type_traits<data_types::i32>::value_type, mem_lock_type::write> lock(instance.output_memory_ptr(2), stream);
+            auto ptr = lock.data();
+            ptr[0] = static_cast<ov::element_type_traits<data_types::i32>::value_type>(valid_input_batch);
+        }
+    }
+}
+}  // namespace
+struct non_max_suppression_gather_impl : typed_primitive_impl<non_max_suppression_gather> {
+    using parent = typed_primitive_impl<non_max_suppression_gather>;
+
+    DECLARE_OBJECT_TYPE_SERIALIZATION(cldnn::cpu::non_max_suppression_gather_impl)
+
+    std::unique_ptr<primitive_impl> clone() const override {
+        return make_unique<non_max_suppression_gather_impl>(*this);
+    }
+
+    non_max_suppression_gather_impl() : parent("non_max_suppression_gather_impl") {}
```

Comment:
[random spot]
Since implementation type is specified as cpu, doesn't this mean that original NMS operation will allocate lockable memory for all outputs? This could potentially affect the performance in that case

Reply:
Fixed to update output memory from input mem, so there is no allocate_memory for outputs.